src/nvim/event/socket.c

```diff
@@ -168,3 +169,64 @@ static void close_cb(uv_handle_t *handle)
     watcher->close_cb(watcher, watcher->data);
   }
 }
+
+static void connect_cb(uv_connect_t *req, int status)
+{
+  int *ret_status = req->data;
+  *ret_status = status;
+}
+
+bool socket_connect(Loop *loop, Stream *stream, bool tcp, const char *address, int timeout)
+{
+  bool success = false;
+  int status;
+  uv_connect_t req;
+  req.data = &status;
+  uv_stream_t *uv_stream;
+  uv_getaddrinfo_t addr_req = {0};
+  const struct addrinfo *addrinfo = NULL;
+  char *addr = NULL;
+  if (tcp) {
+    addr = xstrdup(address);
+    char *host_end = xstrchrnul(addr, ':');
+    if (*host_end != ':') {
+      // TODO: error
+      goto cleanup;
+    }
+    *host_end = NUL;
+
+    const struct addrinfo hints = { .ai_family = AF_UNSPEC, .ai_flags=AI_NUMERICSERV };
+    int retval = uv_getaddrinfo(&loop->uv, &addr_req, NULL, addr, host_end+1, &hints);
+    if (retval != 0) {
+      goto cleanup;
+    }
+    addrinfo = addr_req.addrinfo;
+
+    uv_tcp_t *tcp = &stream->uv.tcp;
+    uv_tcp_init(&loop->uv, tcp);
```

Comment:
Try to move this line after label `tcp_retry` so that `localhost` address works.